design.md

```diff
@@ -0,0 +1,121 @@
+
+### Output streaming
+In order to reduce overhead each command will pipe output to a file named with the a unique identifier tied to the process. This data will be read from the file when requested. We will check to see if the process is still running to determine whether or not to continually read from and stream data from the file, or serve all the output data at once to the client. 
+
+In the cases where processes are still running we will utilize server side streaming from gRPC to serve this data continually to the client until they close the connection or the process finishes.
+
+#### Future Improvements
+* Store output somewhere other than locally on the server, like S3
+* Look into using an external lib that enables tailing from an external source
+* Utilize job management to allow suspending and resuming jobs instead of only being able to start and terminate them
+
+## CLI
+The CLI will utilize the [cobra](https://github.com/spf13/cobra) CLI library.
+
+### Run a process
+The basic structure of the run command will be `run <binary_name> [<args>]`
+
+An example of running a process: `cmdctl run ls -la`
+
+### Get process info
+The basic structure of the get-status and get-output retrieval commands will be: `get-status/get-output [<unique_process_id>]`
+
+An example of querying for process status: `cmdctl get-status shawon-ls-idhhyqm`
+
+We will also have commands for stopping a process and getting the process output. They will follow a similar structure as the above examples.
+
+We will generate a unique identifier for each process, using the username and a generated value, and write the pid to a file named with that id. This way we can track client processes without collisions, while still having something that ties back to the process itself on the host.
+
+### Client Authentication
+The service will utilize mTLS to authenticate and secure requests. For this implementation we will manually create a single local CA certificate that will be used to generate a client side certificate and key pair. The CA will use the `nameConstraints` extension to allow access only to a specific subdomain. In the case of this challenge it will be `permitted:.example.com`. The CA will be trusted by the server to allow client certificates to connect. These certificates and keys will be stored locally for the client.
+
+The client will enforce TLS version 1.3 for maximum security. Each client will need a certificate and key pair generated to interact with the server.
+
+When generating client certificates we will include the the user's email, in the case of this challenge `shawon@example.com`, in the `subjectAltName` field. This will be used as an additional layer of authentication by the CA to ensure only users with an email matching the allowed subdomain can be validated. 
+
+Once the user has been validated their access will be scoped to allow the following on the host via group, which their user will be added to:
```

Comment:
I'm not sure I follow this. Are you saying that you will create an OS-level user for each API user?

Reply:
I have updated the document to include authorizing user access to methods based on email domain. I hope this addresses the authorization issues. 

Ideally long term this would be scoped to users or groups, and would use some form of additional authorization such as a token or oauth flow. For this challenge I felt like making sure that the users are in the domain we want them to be is sufficient.